server/src/main/java/io/crate/expression/operator/DistinctFrom.java

```diff
@@ -87,4 +105,66 @@ public final Boolean evaluate(TransactionContext txnCtx, NodeContext nodeCtx, In
         }
         return argType.compare(arg1, arg2) != 0;
     }
+
+    @Override
+    public Query toQuery(Function function, LuceneQueryBuilder.Context context) {
+        List<Symbol> args = function.arguments();
+        if (!(args.get(0) instanceof Reference ref && args.get(1) instanceof Literal<?> literal)) {
+            return null;
+        }
+        String storageIdentifier = ref.storageIdent();
+        Object value = literal.value();
+        if (value == null) {
+            if (!ref.isNullable()) {
+                // If the column is NOT NULL, `x IS DISTINCT FROM NULL` is true for all documents
+                return new MatchAllDocsQuery();
+            }
+            return IsNullPredicate.refExistsQuery(ref, context);
+        }
+
+        DataType<?> dataType = ref.valueType();
+        return switch (dataType.id()) {
+            case ObjectType.ID -> {
+                var query = EqOperator.refEqObject(
+                    function,
+                    ref.column(),
+                    (ObjectType) dataType,
+                    (Map<String, Object>) value,
+                    context,
+                    BooleanClause.Occur.MUST_NOT
+                );
+                if (query instanceof GenericFunctionQuery) {
+                    yield query;
+                }
+                if (query instanceof BooleanQuery bq) {
+                    if (bq.clauses().size() != 1) {
+                        yield query;
+                    }
+                    var boolQuery = new BooleanQuery.Builder();
+                    boolQuery.add(bq.clauses().getFirst().getQuery(), BooleanClause.Occur.MUST_NOT);
+                    boolQuery.add(genericFunctionFilter(function, context), BooleanClause.Occur.FILTER);
+                    yield boolQuery.build();
+                }
+                yield query;
```

Comment:
I don't really understand why this is necessary here. Couldn't `refEqObject` just do the right thing directly?

Seems potentially error prone to rely on the clauses size to determine if it needs a genericFunctionFilter or not.

Reply:
Yes right, not perfect. The main problem here again is that we don't want the optional generic function filter to be negated.

Reply:
Wouldn't it be possible to adapt the following: 

https://github.com/crate/crate/blob/6a90ae186cec2da9f9a7455a28fd42c8fdbaef0d/server/src/main/java/io/crate/expression/operator/EqOperator.java#L334-L340

too by using the new `termQueryOccur` parameter? 

And I think it deserves a comment to highlight that were mixing `!=` and `=` semantics.
Or could we just create a `Eq` `Function` instance based on the ref, and literal, and use that + always wrap in `not` ?

(Which makes me think, the benchmark results might be due to double query negation with a `is not distinct from` - I think it's good enough for now but could be followed up on to verify and maybe mitigate)

Reply:
I'm not sure if I understand your suggestion as we must negate the terms query and in your example it's only about how to add the generic filter.
But I found a maybe better approach, see my fixup commit. This also eliminates the need to a generic filter for known child cases.
What I don't completely understand is, why negating the query using `Query.not` which results in a boolean query combining `MatchAll` + the terms query works, while adding the terms query into a boolean clause using the `MUST_NOT` occur won't.

Reply:
The benchmark results of my recent changes (fixup) didn't change (expected, as I didn't touch the primitive code path).